src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,387 @@
+    pub frp : Frp
+}
+
+impl Breadcrumbs {
+    /// Constructor.
+    pub fn new<'t,S:Into<&'t Scene>>(scene:S, focus_manager:&FocusManager) -> Self {
+        let frp     = Frp::new();
+        let model   = Rc::new(BreadcrumbsModel::new(scene,&frp,focus_manager));
+        let network = &frp.network;
+
+
+        // === Breadcrumb selection ===
+
+        let update_selection = f!([model]((old_index,new_index)) {
+            if old_index > 0 {
+                model.breadcrumbs.borrow_mut().get(old_index-1).map(|breadcrumb:&Breadcrumb| {
+                    breadcrumb.frp.deselect.emit(())
+                });
+            }
+            if new_index > 0 {
+                model.breadcrumbs.borrow_mut().get(new_index-1).map(|breadcrumb:&Breadcrumb| {
+                    breadcrumb.frp.select.emit(())
+                });
+            }
+            (old_index,new_index)
+        });
+
+        let update_selection_clone = update_selection.clone();
+        frp::extend! { network
+            _push_breadcrumb <- frp.push_breadcrumb.map(f!([model](local_call) {
+                let indices = model.push_breadcrumb(local_call);
+                indices.map(&update_selection_clone).map(|(_,new_index)| {
+                    if new_index > 0 {
+                        model.breadcrumbs.borrow_mut().get(new_index-1).map(|breadcrumb| {
+                            breadcrumb.frp.fade_in.emit(());
+                        });
+                    }
+                });
+            }));
+            _pop_breadcrumb <- frp.pop_breadcrumb.map(f!([model](_) {
+                model.pop_breadcrumb().map(&update_selection);
+            }));
+        }
+
+
+        // === Project Name ===
+
+        frp::extend! { network
+            eval frp.project_name((name) model.project_name.frp.name.emit(name));
+            frp.outputs.project_name <+ model.project_name.frp.outputs.name;
+        }
+
+
+        // === GUI Update ===
+
+        frp::extend! { network
+            eval model.project_name.frp.outputs.width((width) {
+                model.relayout_for_project_name_width(*width)
+            });
+        }
+
+
+        // === User Interaction ===
+
+        frp::extend! {network
+            _mouse_down <- model.project_name.frp.outputs.mouse_down.map(f!([model] (_) {
+                let (popped_count,local_calls) = model.select_breadcrumb(0);
+                for _ in 0..popped_count {
+                    model.frp_outputs.breadcrumb_pop.emit(());
+                }
+                for local_call in local_calls {
+                    model.frp_outputs.breadcrumb_push.emit(local_call);
+                }
+            }));
+            eval_ frp.cancel_project_name_editing(model.project_name.frp.cancel_editing.emit(()));
+            commit_if_changed <- frp.outside_press.gate(&model.project_name.frp.outputs.edit_mode);
+            eval_ commit_if_changed({
+                model.project_name.frp.commit.emit(())
+            });
```

Comment:
We have here the same problem we had earlier in graph_editor. After evaluating `outside_press`, breadcrumbs are checking an internal state of `project_name` and are then controlling its functions basedo n its internal state. It is not responsibility of this object to check the internal state. Instead, this should just call `project_name.frp.commit_if_changed` and it should be further delegated. Basically, each object knows its internal state, and its not the responsibility of the parent to check that state and make decisions based on it.